File: _archived/notifications-nodejs/src/integrations/providers/adapters/sendgrid.adapter.ts
import * as https from "https";
import { EmailProviderAdapter, EmailSendPayload, EmailSendResult } from "../interfaces/email-provider.interface";
import { FailureCategory } from "../../../types";
import { logger } from "../../../shared/logger";

interface SendGridConfig {
  apiKey: string;
  defaultFromEmail: string;
  defaultFromName: string;
}

function classifySendGridError(statusCode: number, body: string): FailureCategory {
  if (statusCode === 401 || statusCode === 403) return "auth_config_failure";
  if (statusCode === 400) {
    const lower = body.toLowerCase();
    if (lower.includes("invalid") && (lower.includes("email") || lower.includes("recipient"))) {
      return "invalid_recipient";
    }
    return "non_retryable_failure";
  }
  if (statusCode === 413 || statusCode === 422) return "non_retryable_failure";
  if (statusCode === 429 || statusCode >= 500) return "retryable_provider_failure";
  return "non_retryable_failure";
}

function httpsPost(url: string, headers: Record<string, string>, body: string): Promise<{ statusCode: number; body: string }> {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const options = {
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      method: "POST",
      headers: { ...headers, "Content-Length": Buffer.byteLength(body) },
    };

    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => { data += chunk; });
      res.on("end", () => resolve({ statusCode: res.statusCode ?? 0, body: data }));
    });

    req.on("error", reject);
    req.setTimeout(10000, () => { req.destroy(new Error("Request timed out")); });
    req.write(body);
    req.end();
  });
}

function httpsGet(url: string, headers: Record<string, string>): Promise<{ statusCode: number; body: string }> {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const options = {
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      method: "GET",
      headers,
    };

    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => { data += chunk; });
      res.on("end", () => resolve({ statusCode: res.statusCode ?? 0, body: data }));
    });

    req.on("error", reject);
    req.setTimeout(8000, () => { req.destroy(new Error("Health check timed out")); });
    req.end();
  });
}

export class SendGridEmailProviderAdapter implements EmailProviderAdapter {
  readonly providerType = "sendgrid";
  private config: SendGridConfig;

  constructor(config: SendGridConfig) {
    this.config = config;
  }

  async validateConfig(): Promise<boolean> {
    const { apiKey, defaultFromEmail } = this.config;
    if (!apiKey || !defaultFromEmail) {
      logger.warn("SendGrid config is incomplete", { hasApiKey: !!apiKey, hasFromEmail: !!defaultFromEmail });
      return false;
    }
    return true;
  }

  async send(payload: EmailSendPayload): Promise<EmailSendResult> {
    const valid = await this.validateConfig();
    if (!valid) {
      return {
        success: false,
        failure: {
          category: "auth_config_failure",
          message: "SendGrid is not configured — SENDGRID_API_KEY or SENDGRID_DEFAULT_FROM_EMAIL is missing",
          retryable: false,
        },
      };
    }

    const body = JSON.stringify({
      personalizations: [{ to: [{ email: payload.to }] }],
      from: {
        email: payload.from || this.config.defaultFromEmail,
        name: this.config.defaultFromName || undefined,
      },
      subject: payload.subject,
      content: [
        { type: "text/plain", value: payload.body },
        ...(payload.html ? [{ type: "text/html", value: payload.html }] : []),
      ],
    });

    try {
      const result = await httpsPost(
        "https://api.sendgrid.com/v3/mail/send",
        {
          Authorization: `Bearer ${this.config.apiKey}`,
          "Content-Type": "application/json",
        },
        body
      );

      if (result.statusCode === 202) {
        logger.info("SendGrid: email sent successfully", { to: payload.to });
        return { success: true };
      }

      const category = classifySendGridError(result.statusCode, result.body);
      logger.warn("SendGrid: send failed", { statusCode: result.statusCode, category });

      return {
        success: false,
        failure: {
          category,
          providerCode: String(result.statusCode),
          message: result.body.slice(0, 500),
          retryable: category === "retryable_provider_failure",
        },
      };
    } catch (err) {
      const message = String(err);
      const isTimeout = message.includes("timed out");
      logger.error("SendGrid: network error during send", { error: message });
      return {
        success: false,
        failure: {
          category: isTimeout ? "provider_unavailable" : "retryable_provider_failure",
          message,
          retryable: true,
        },
      };
    }
  }

  /**
   * Queries the SendGrid Email Activity API for the latest delivery status of a recently
   * sent message.  Returns null when the message hasn't appeared yet or when the API
   * isn't accessible (e.g. plan doesn't include Email Activity).
   */
  async queryMessageStatus(
    toEmail: string,
  ): Promise<"delivered" | "not_delivered" | "blocked" | "bounced" | "deferred" | "spam_report" | "processing" | null> {
    if (!this.config.apiKey) return null;
    try {
      const query = encodeURIComponent(`to_email="${toEmail}"`);
      const result = await httpsGet(
        `https://api.sendgrid.com/v3/messages?query=${query}&limit=1&orderby=last_event_time+desc`,
        { Authorization: `Bearer ${this.config.apiKey}` },
      );
      if (result.statusCode !== 200) {
        logger.warn("SendGrid Messages API returned non-200", { statusCode: result.statusCode, body: result.body.slice(0, 200) });
        return null;
      }
      const body = JSON.parse(result.body) as { messages?: { status: string }[] };
      const first = body.messages?.[0];
      if (!first) return null;
      const s = first.status as string;
      const validStatuses = ["delivered", "not_delivered", "blocked", "bounced", "deferred", "spam_report", "processing"];
      if (validStatuses.includes(s)) {
        return s as "delivered" | "not_delivered" | "blocked" | "bounced" | "deferred" | "spam_report" | "processing";
      }
      return null;
    } catch (err) {
      logger.warn("SendGrid queryMessageStatus error", { error: String(err) });
      return null;
    }
  }

  async healthCheck(): Promise<{ status: "healthy" | "degraded" | "down"; latencyMs?: number }> {
    if (!this.config.apiKey) {
      return { status: "down" };
    }

    const start = Date.now();
    try {
      const result = await httpsGet("https://api.sendgrid.com/v3/scopes", {
        Authorization: `Bearer ${this.config.apiKey}`,
      });
      const latencyMs = Date.now() - start;

      if (result.statusCode === 200) return { status: "healthy", latencyMs };
      if (result.statusCode === 401 || result.statusCode === 403) return { status: "down", latencyMs };
      if (result.statusCode >= 500) return { status: "degraded", latencyMs };
      return { status: "healthy", latencyMs };
    } catch {
      return { status: "down", latencyMs: Date.now() - start };
    }
  }
}

File: apps/services/notifications/Notifications.Infrastructure/Providers/Adapters/SendGridAdapter.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using Notifications.Application.Interfaces;

namespace Notifications.Infrastructure.Providers.Adapters;

public class SendGridAdapter : IEmailProviderAdapter
{
    public string ProviderType => "sendgrid";

    private readonly string _apiKey;
    private readonly string _defaultFromEmail;
    private readonly string _defaultFromName;
    private readonly HttpClient _http;
    private readonly ILogger<SendGridAdapter> _logger;

    public SendGridAdapter(string apiKey, string defaultFromEmail, string defaultFromName, HttpClient http, ILogger<SendGridAdapter> logger)
    {
        _apiKey = apiKey;
        _defaultFromEmail = defaultFromEmail;
        _defaultFromName = defaultFromName;
        _http = http;
        _logger = logger;
    }

    public Task<bool> ValidateConfigAsync()
        => Task.FromResult(!string.IsNullOrEmpty(_apiKey) && !string.IsNullOrEmpty(_defaultFromEmail));

    public async Task<EmailSendResult> SendAsync(EmailSendPayload payload)
    {
        if (!await ValidateConfigAsync())
            return new EmailSendResult { Success = false, Failure = new ProviderFailure { Category = "auth_config_failure", Message = "SendGrid is not configured", Retryable = false } };

        var body = new
        {
            personalizations = new[] { new { to = new[] { new { email = payload.To } } } },
            from = new { email = payload.From ?? _defaultFromEmail, name = _defaultFromName },
            subject = payload.Subject,
            content = new List<object> { new { type = "text/plain", value = payload.Body } }
                .Concat(payload.Html != null ? new[] { new { type = "text/html", value = payload.Html } } : Array.Empty<object>()).ToArray()
        };

        var request = new HttpRequestMessage(HttpMethod.Post, "https://api.sendgrid.com/v3/mail/send");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
        request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
            var response = await _http.SendAsync(request, cts.Token);
            var statusCode = (int)response.StatusCode;
            var responseBody = await response.Content.ReadAsStringAsync(cts.Token);

            if (statusCode == 202)
            {
                // Extract the provider message ID from the X-Message-Id response header.
                // This ID is required to correlate incoming webhook events (delivered,
                // bounced, opened, etc.) back to the correct NotificationAttempt record.
                response.Headers.TryGetValues("X-Message-Id", out var msgIdValues);
                var providerMessageId = msgIdValues?.FirstOrDefault();
                _logger.LogInformation("SendGrid: email sent successfully to {To}, messageId={MessageId}", payload.To, providerMessageId ?? "none");
                return new EmailSendResult { Success = true, ProviderMessageId = providerMessageId };
            }

            var category = ClassifyError(statusCode, responseBody);
            _logger.LogWarning("SendGrid: send failed {StatusCode} {Category}", statusCode, category);
            return new EmailSendResult { Success = false, Failure = new ProviderFailure { Category = category, ProviderCode = statusCode.ToString(), Message = responseBody[..Math.Min(responseBody.Length, 500)], Retryable = category == "retryable_provider_failure" } };
        }
        catch (Exception ex)
        {
            var isTimeout = ex is TaskCanceledException or OperationCanceledException;
            _logger.LogError(ex, "SendGrid: network error during send");
            return new EmailSendResult { Success = false, Failure = new ProviderFailure { Category = isTimeout ? "provider_unavailable" : "retryable_provider_failure", Message = ex.Message, Retryable = true } };
        }
    }

    public async Task<ProviderHealthResult> HealthCheckAsync()
    {
        if (string.IsNullOrEmpty(_apiKey)) return new ProviderHealthResult { Status = "down" };
        var start = DateTime.UtcNow;
        try
        {
            var request = new HttpRequestMessage(HttpMethod.Get, "https://api.sendgrid.com/v3/scopes");
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(8));
            var response = await _http.SendAsync(request, cts.Token);
            var latencyMs = (int)(DateTime.UtcNow - start).TotalMilliseconds;
            var status = (int)response.StatusCode switch
            {
                200 => "healthy",
                401 or 403 => "down",
                >= 500 => "degraded",
                _ => "healthy"
            };
            return new ProviderHealthResult { Status = status, LatencyMs = latencyMs };
        }
        catch
        {
            return new ProviderHealthResult { Status = "down", LatencyMs = (int)(DateTime.UtcNow - start).TotalMilliseconds };
        }
    }

    private static string ClassifyError(int statusCode, string body)
    {
        if (statusCode is 401 or 403) return "auth_config_failure";
        if (statusCode == 400)
        {
            var lower = body.ToLowerInvariant();
            if (lower.Contains("invalid") && (lower.Contains("email") || lower.Contains("recipient"))) return "invalid_recipient";
            return "non_retryable_failure";
        }
        if (statusCode is 413 or 422) return "non_retryable_failure";
        if (statusCode is 429 or >= 500) return "retryable_provider_failure";
        return "non_retryable_failure";
    }
}
